### PR TITLE
Make sure that `conda env update -f` sets env vars from the referenced yaml file

### PR DESCRIPTION
**Pull request by FaustinCarter**
Assume an environment file called `foo.yaml` that looks like:

```yaml
name: foo
dependencies:
    - python
variables:
    MY_VAR: bar
```

When the user runs `conda env create -f foo.yaml`, conda will create an environment named `foo`, install Python in that environment, and set an env var `MY_VAR=bar`.

Now imagine the user modifies the file:

```yaml
name: foo
dependencies:
    - python
    - ipython
variables:
    MY_VAR: bar
    MY_VAR2: baz
```

And then runs: `conda env update -f foo.yaml`. The expectation is that IPython would be installed, and `MY_VAR2` would be set to `baz`. However, currently the `conda env update` command ignores any env variables specified in the yaml file. This PR attempts to remedy that.

Potential issue: This doesn't attempt to track env vars that are removed from the file.

### CODE REVIEW

**Comment by matt2000** [comment]
Regarding:

> Potential issue: This doesn't attempt to track env vars that are removed from the file.

I think this is OK. There's no way to know what the user expects of their environment after they remove a var defined here. I.e., it would be wrong to unset a missing environment var; the user may want the current value, and only intend for conda not to set it any more. If a user intended to unset a variable in their current environment, they'll have to do that themselves when they remove it from their environment file, and I think most users would expect this.

**Comment by jezdez** [comment]
@FaustinCarter Yeah, it seems like a reasonable idea that updating the env vars in environment files also persists them on update and reduce the need to recreate the environment from scratch or find other ways to update them.

FWIW this would be a good case to test, maybe here: https://github.com/conda/conda/blob/ca4069a937325c6f0279b72859fb87b7b8939478/tests/conda_env/test_create.py#L68 

I think this just needs to check somehow if the environment variables are persisteted on update.

Do you have cycles to work on the tests?

**Review by jezdez — CHANGES_REQUESTED** [review]
This just needs some tests to make sure the new behavior is going to be checked going forward. I see a lot of duplicate code between the `env create` and `env update` commands, so this is ripe for a refactor at some point.

**Comment by jezdez** [comment]
#10448 previously reported that as well

**Review by magau — APPROVED** [review]
I totally agree with  @jezdez comment:
> I see a lot of duplicate code between the `env create` and `env update` commands, so this is ripe for a refactor at some point.

However, as a fast fix for this issue, the proposed changes works fine.
I've already tested it in the PR #10449, that I've made by myself, 5 month ago proposing exactly the same changes, that was closed in favor of this one...

(Unfortunately, mine was rejected, due to a missing PDF signature, that I didn't sign because I don't know you from now where.)

Hope it will be approved this time ;)

Thanks all of you that contibut for this wonderful environment manager (conda) that makes our lives so much easyer in a way that it keeps platform and versions independent code working on!

**Comment by jezdez** [comment]
> I totally agree with @jezdez comment:
> 
> > I see a lot of duplicate code between the `env create` and `env update` commands, so this is ripe for a refactor at some point.
> 
> However, as a fast fix for this issue, the proposed changes works fine.
> I've already tested it in the PR #10449, that I've made by myself 5 month ago proposing exactly the same changes, that was closed in favor of this one...

Sorry, to clarify what I mean with tests, I mean extending the conda test suite to test for the behavior that you're expecting, not just a manual test.

> (Unfortunately, mine was rejected, due to a missing PDF signature, that I didn't sign because I don't know you from no where.)

The PR wasn't rejected because you didn't sign the CLA but because this pull request exists and its author has signed the CLA. Unfortunately signing it is a legal requirement to add 3rd party code contributions to the conda code base. It's a regularl practice in permissively licensed Open Source projects. 

> Hope it will be approved this time ;)

If you have the spare cycles to extend the test suite, it'll be approved right away.

> Thanks all of you that contibut for this wonderful environment manager (conda) that makes our lives so much easyer in a way that it keeps platform and versions independent code working on!

Thanks!

**Comment by FaustinCarter** [comment]
I'll try and take a stab at adding some tests in the near future. I haven't forgotten about this.

**Comment by FaustinCarter** [comment]
@jezdez Can you unblock this or update the tags so tests can run? Thanks,

**Comment by jezdez** [comment]
> @jezdez Can you unblock this or update the tags so tests can run? Thanks,

Done, this was related to GitHub's anti-cryptominer policy for new contributors.

**Comment by FaustinCarter** [comment]
@jezdez OK, I made a dumb mistake. I've fixed it. Three things to note:
1. I was checking `os.environ` for vars, which was failing because the env never activates during test. I've switched to checking the env vars via `PrefixData` instead.
2. For some reason, Linux tests were marked as passing _even though my tests failed_!
3. The `test_create_update_remote_env_file` will fail until this is merged because it uses the environment file on the github master branch (which I update in this PR, so that test won't pass until this is merged).

**Comment by FaustinCarter** [comment]
@jezdez This is ready to go. The only test that is failing is expected to fail unless we open another PR to update the yaml files in master ahead of this PR. I still don't know why the linux jobs are all reading as passing when they all have a single test failure. That seems broken (although unrelated to this PR).

**Comment by chenghlee** [comment]
Moving this to "Release TBD" so we can bundle it into a bigger release that specifically targets `conda env`-release issues.

**Comment by shane-axiom** [comment]
Any way to get this into a release? Ran into this today and while I'm happy to see a fix waiting in the wings it's disappointing to see it stalled at the finish line for five months ;)

**Comment by jezdez** [comment]
reopening to kickstart CI..

**Comment by jezdez** [comment]
@FaustinCarter I'm sorry but something is wrong with the tests

**Comment by FaustinCarter** [comment]
@ jezdez 

> @FaustinCarter I'm sorry but something is wrong with the tests

@jezdez Did you see my earlier note in the thread about how the tests are pulling a yaml file from master that this PR modifies?

**Comment by kenodegard** [comment]
@FaustinCarter @jezdez

Yes, I agree that the 6 CI tests failing for `IntegrationTests.test_create_update_remote_env_file` are expected due to the test fetching the file from the master branch url instead of the local filesystem (see `tests.conda_env.support_file`). Since the `IntegrationTests.test_create_update` tests pass I agree it would be acceptable to ignore those failures.

The failing pre-commit check seems to be a matter of rebasing these changes in order to get the correct pre-commit config included (we recently enabled pre-commit checks on the repo). Its not liking the missing blank EOF for the yaml files.

No idea why readthedocs is failing.

**Comment by FaustinCarter** [comment]
I rebased off of master and pushed up since it's been so long. Not sure if that was the right call or not since now a pre-commit check is failing.

**Comment by kenodegard** [comment]
@FaustinCarter That rebase seemed to have gone a little haywire, I went ahead and fixed the issue. This will be ready to merge once the tests complete. Thanks!

**Comment by jezdez** [comment]
> @ jezdez
> 
> > @FaustinCarter I'm sorry but something is wrong with the tests
> 
> @jezdez Did you see my earlier note in the thread about how the tests are pulling a yaml file from master that this PR modifies?

Oh, I'm sorry that I missed this, hadn't seen the `support_file` utility before, thanks for bringing this up again!

I'm inclined to say that the `support_file` function shouldn't even have the remote option and should instead start a local webserver to fetch the file from there, to have the tests decoupled from the master branch and GitHub as a service in the first place. But that's topic for another PR and shouldn't block this one. Thanks again for raising this!

**Review by jezdez — APPROVED** [review]
Looks good to me and definitely warrants a new issue to make sure we stop using `support_file` in a way that depends on the master branch on GitHub.

**Pull request merged**
